pilot/pkg/model/mesh.go

```diff
@@ -0,0 +1,785 @@
+
+	// PORT represents the recipient port on the server IP address, Example: 443
+	PORT destinationAttribute = "destination.port"
+
+	// DOMAIN represents the domain portion of the service name, excluding
+	// the name and namespace, example: svc.cluster.local
+	DOMAIN destinationAttribute = "destination.domain"
+
+	// USER represents the __CANONICAL__ Mesh user running the destination
+	// application, example: service-account-foo. The user's identity is
+	// assumed to be trusted within the entire mesh, irrespective of which
+	// platform, cluster or availability zone this instance runs in.
+	USER destinationAttribute = "destination.user"
+)
+
+var (
+	destinationAttrSet = map[string]bool{
+		UID.stringValue():       true,
+		SERVICE.stringValue():   true,
+		NAME.stringValue():      true,
+		NAMESPACE.stringValue(): true,
+		IP.stringValue():        true,
+		PORT.stringValue():      true,
+		DOMAIN.stringValue():    true,
+		USER.stringValue():      true,
+	}
+
+	canonicalIstioDomain = "svc.cluster.local"
+	svcRegexp            = regexp.MustCompile(
+		"([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.svc\\.cluster\\.local")
+	domainRegexp = regexp.MustCompile(
+		"([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])*")
+)
+
+type destinationAttribute string
+
+// Mesh is a platform independent abstraction used by Controllers
+// for maintaining a list of service endpoints used by this Pilot.
+//
+// Controllers under https://github.com/istio/istio/tree/master/pilot/platform
+// update the list of service endpoints in this Mesh with those
+// available in a platform specific service registry.
+//
+// Under the hoods, the Mesh implements the necessary logic required for endpoint
+// discovery based on routing rules and endpoint subsets.
+// See https://github.com/istio/api/search?q=in%3Afile+"message+Subset"+language%3Aproto
+// This logic includes comparing the updated list provided by Controllers with
+// what this view holds and accordingly updating internal structures used for
+// endpoint discovery.
+type Mesh struct {
+
+	// allEndpoints is a map of the UID to the server instance. The UID is expected
+	// to be unique across the mesh. Mesh uses this map during updates from
+	// Controllers to locate the previous representation  of the endpoint and check
+	// for changes to its metadata that could result in changes to the subsetEndpoints.
+	allEndpoints map[string]*Endpoint
+
+	// subsetDefinitions holds the metadata, basically the set of lable name values
+	// that ought to match corresponding attributes of the endpoint.
+	// The actual subset of allEndpoints that match these subset definitions is
+	// stored in subsetEndpoints. Mesh uses this map for identifying config changes
+	// to the subset's metadata. The key of this map is the Subset name.
+	subsetDefinitions map[string]*route.Subset
```

Comment:
subset names are not unique across services. Its only unique within a destination rule/within a service.

Reply:
I believe it's supposed to be a qualified name - the assumption in the design was that all rules are somehow synchronized and visible in the entire mesh. Example key/values would really help.

Reply:
What is the API for exchanging rules (subset names)?

Reply:
We decided to keep this simple - how the rules are exchanged is a separate problem, left to CICD or scripts or whatever the admin wants. We assume they'll make it to each apiserver. Maybe galley or someone else will add extra automation - but not a problem for this PR.